File: src/de/ax/powermode/PowerMode.scala
/*
 * Copyright 2015 Baptiste Mesta
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package de.ax.powermode

import java.awt.event.InputEvent
import java.io.File
import javax.swing.KeyStroke
import com.intellij.openapi.Disposable
import com.intellij.openapi.actionSystem.DataContext
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.components.{ApplicationComponent, PersistentStateComponent, State, Storage}
import com.intellij.openapi.editor.EditorFactory
import com.intellij.openapi.editor.actionSystem.EditorActionManager
import com.intellij.util.xmlb.XmlSerializerUtil
import de.ax.powermode.power.management.ElementOfPowerContainerManager
import org.apache.log4j._
import org.jetbrains.annotations.Nullable

import scala.collection.immutable.Seq
import scala.util.Try

/**
  * @author Baptiste Mesta
  */
object PowerMode {

  val logger: Logger = Logger.getLogger(classOf[PowerMode])

  @Nullable def getInstance: PowerMode = {
    try {
      ApplicationManager.getApplication.getComponent(classOf[PowerMode])
    } catch {
      case e: Throwable =>
        logger.debug("error getting component: " + e.getMessage(), e)
        null
    }
  }


  def obtainColorEdges(pm: PowerMode): ColorEdges = {
    import pm._
    val edges = new ColorEdges()
    edges.setAlpha(getColorAlpha)
    edges.setRedFrom(getRedFrom)
    edges.setRedTo(getRedTo)
    edges.setGreenFrom(getGreenFrom)
    edges.setGreenTo(getGreenTo)
    edges.setBlueFrom(getBlueFrom)
    edges.setBlueTo(getBlueTo)
    edges
  }
}

@State(name = "PowerModeII", storages = Array(new Storage(file = "$APP_CONFIG$/power.mode.ii.xml")))
class PowerMode extends ApplicationComponent with PersistentStateComponent[PowerMode] {
  var bamLife: Long = 1000

  var soundsFolder = Option.empty[File]


  var gravityFactor: Double = 21.21

  var sparkVelocityFactor: Double = 4.36

  var sparkSize = 3

  var sparksEnabled = true

  var frameRate: Int = 30

  def setFrameRate(f: Int) {
    frameRate = f
  }

  def getFrameRate() = frameRate

  var maxFlameSize = 100

  var maxFlameLife = 2000

  var heatupTime = 10000

  type HeatupKey = (Option[KeyStroke], Long)

  var lastKeys = List.empty[HeatupKey]

  var keyStrokesPerMinute = 300
  var heatupFactor = 1.0
  var sparkLife = 3000
  var sparkCount = 10
  var shakeRange = 4
  var flamesEnabled: Boolean = true
  var maybeElementOfPowerContainerManager = Option.empty[ElementOfPowerContainerManager]
  private var enabled: Boolean = true
  private var shakeEnabled: Boolean = true
  var isBamEnabled: Boolean = true
  var isSoundsPlaying = false
  var powerIndicatorEnabled = true

  def flameImageFolder = {
    if (!_isCustomFlameImages) Some(new File("fire/animated/256")) else customFlameImageFolder
  }

  def bamImageFolder = {
    if (!_isCustomBamImages) Some(new File("bam")) else customBamImageFolder
  }

  def increaseHeatup(dataContext: Option[DataContext] = Option.empty[DataContext], keyStroke: Option[KeyStroke] = Option.empty[KeyStroke]): Unit = {
    val ct = System.currentTimeMillis()
    lastKeys = (keyStroke, ct) :: filterLastKeys(ct)
    dataContext.foreach(dc => maybeElementOfPowerContainerManager.foreach(_.showIndicator(dc)))

  }


  def reduceHeatup: Unit = {
    val ct = System.currentTimeMillis()
    lastKeys = filterLastKeys(ct)
    //    maybeElementOfPowerContainerManager.map(_.showIndicator)
  }

  private def filterLastKeys(ct: Long): List[HeatupKey] = {
    lastKeys.filter(_._2 >= ct - heatupTime)
  }

  def rawValueFactor = {
    val base = heatupFactor +
      ((1 - heatupFactor) * rawTimeFactor)
    val elems = (base - heatupThreshold) / (1 - heatupThreshold)
    elems
  }

  def valueFactor: Double = {
    val base = heatupFactor +
      ((1 - heatupFactor) * timeFactor)
    val elems = (base - heatupThreshold) / (1 - heatupThreshold)
    elems
    val max = Seq(elems, 0.0).max
    assert(max <= 1)
    assert(max >= 0)
    max
  }


  var hotkeyWeight: Double = keyStrokesPerMinute * 0.05

  def rawTimeFactor: Double = {
    val tf = Try {
      if (heatupTime < 1000) {
        1
      } else {
        val d = heatupTime.toDouble / (60000.0 / keyStrokesPerMinute)
        val keysWorth = lastKeys.map {
          case (Some(ks), _) =>
            val size = Seq(InputEvent.CTRL_DOWN_MASK, InputEvent.ALT_DOWN_MASK, InputEvent.SHIFT_DOWN_MASK).count(m => (ks.getModifiers & m) > 0)
            val res = size * hotkeyWeight
            res
          case _ => 1
        }.sum
        keysWorth / d
      }
    }.getOrElse(0.0)
    tf
  }


  def timeFactor: Double = {
    val tf = Try {
      if (heatupTime < 1000) {
        1
      } else {
        val d = heatupTime.toDouble / (60000.0 / keyStrokesPerMinute)
        val keysWorth = lastKeys.map {
          case (Some(ks), _) =>
            val size = Seq(InputEvent.CTRL_DOWN_MASK, InputEvent.ALT_DOWN_MASK, InputEvent.SHIFT_DOWN_MASK).count(m => (ks.getModifiers & m) > 0)
            val res = size * hotkeyWeight
            res
          case _ => 1
        }.sum
        math.min(keysWorth, d) / d
      }
    }.getOrElse(0.0)
    tf
  }


  var caretAction: Boolean = true

  override def initComponent: Unit = {
    val editorFactory = EditorFactory.getInstance
    maybeElementOfPowerContainerManager = Some(new ElementOfPowerContainerManager)
    maybeElementOfPowerContainerManager.foreach(editorFactory.addEditorFactoryListener(_, new Disposable() {
      def dispose {
      }
    }))
    val editorActionManager = EditorActionManager.getInstance
    EditorFactory.getInstance().getEventMulticaster.addCaretListener(new MyCaretListener())
    maybeElementOfPowerContainerManager.map(cm =>
      editorActionManager.getTypedAction.setupRawHandler
      (new MyTypedActionHandler(editorActionManager.getTypedAction.getRawHandler)))
  }

  override def disposeComponent: Unit = {
    maybeElementOfPowerContainerManager.foreach(_.dispose)
  }

  override def getComponentName: String = {
    return "PowerModeII"
  }

  def getState: PowerMode = {
    return this
  }

  def loadState(state: PowerMode) {
    XmlSerializerUtil.copyBean(state, this)
  }

  def isEnabled: Boolean = {
    enabled
  }

  def setEnabled(enabled: Boolean) {
    this.enabled = enabled
  }

  def isShakeEnabled: Boolean = {
    shakeEnabled
  }

  def setShakeEnabled(shakeEnabled: Boolean) {
    this.shakeEnabled = shakeEnabled
  }

  def getSparkCount = sparkCount

  def setSparkCount(sparkCount: Int) {
    this.sparkCount = sparkCount
  }

  def getSparkLife = sparkLife

  def setSparkLife(sparkRange: Int) {
    this.sparkLife = sparkRange
  }

  def getShakeRange = shakeRange

  def setShakeRange(shakeRange: Int) {
    this.shakeRange = shakeRange
  }

  def getHeatup = (heatupFactor * 100).toInt

  def setHeatup(heatup: Int) {
    this.heatupFactor = heatup / 100.0
  }

  def getHeatupTime = heatupTime

  def setHeatupTime(heatupTime: Int) {
    this.heatupTime = math.max(0, heatupTime)
  }

  def getFlameLife: Int = {
    return maxFlameLife
  }

  def setFlameLife(flameLife: Int): Unit = {
    maxFlameLife = flameLife
  }

  def getmaxFlameSize: Int = {
    return maxFlameSize
  }

  def setmaxFlameSize(maxFlameSize: Int): Unit = {
    this.maxFlameSize = maxFlameSize
  }

  def getKeyStrokesPerMinute: Int = {
    return keyStrokesPerMinute
  }

  def setKeyStrokesPerMinute(keyStrokesPerMinute: Int) {
    this.keyStrokesPerMinute = keyStrokesPerMinute
  }

  def isFlamesEnabled: Boolean = {
    return flamesEnabled
  }

  def setFlamesEnabled(flamesEnabled: Boolean) {
    this.flamesEnabled = flamesEnabled
  }

  def isSparksEnabled: Boolean = {
    return sparksEnabled
  }

  def setSparksEnabled(sparksEnabled: Boolean) {
    this.sparksEnabled = sparksEnabled
  }

  def getSparkSize: Int = {
    return sparkSize
  }

  def setSparkSize(sparkSize: Int) {
    this.sparkSize = sparkSize
  }

  def getGravityFactor(): Double = gravityFactor

  def setGravityFactor(f: Double) {
    gravityFactor = f
  }

  def getSparkVelocityFactor(): Double = sparkVelocityFactor

  def setSparkVelocityFactor(f: Double) {
    sparkVelocityFactor = f
  }

  var redFrom: Int = 200

  def getRedFrom: Int = {
    redFrom
  }

  var redTo: Int = 255

  def getRedTo: Int = {
    return redTo
  }

  var greenTo: Int = 255

  def getGreenTo: Int = {
    return greenTo
  }

  var greenFrom: Int = 0

  def getGreenFrom: Int = {
    return greenFrom
  }

  var blueFrom: Int = 0

  def getBlueFrom: Int = {
    return blueFrom
  }

  var blueTo: Int = 103

  def getBlueTo: Int = {
    return blueTo
  }

  var colorAlpha: Int = 164

  def getColorAlpha: Int = {
    return colorAlpha
  }

  def setRedFrom(redFrom: Int) {
    if (redFrom <= redTo)
      this.redFrom = redFrom
  }

  def setRedTo(redTo: Int) {
    if (redTo >= redFrom)
      this.redTo = redTo
  }

  def setGreenFrom(gf: Int) {
    if (gf <= greenTo)
      greenFrom = gf
  }

  def setGreenTo(greenTo: Int) {
    if (greenTo >= greenFrom)
      this.greenTo = greenTo
  }

  def setBlueFrom(bf: Int) {
    if (bf <= blueTo)
      blueFrom = bf
  }

  def setBlueTo(blueTo: Int) {
    if (blueTo >= getBlueFrom)
      this.blueTo = blueTo
  }

  def setColorAlpha(alpha: Int) {
    colorAlpha = alpha
  }

  def setSoundsFolder(file: String) {
    soundsFolder = Option(new File(file))
  }

  def getSoundsFolder = soundsFolder.map(_.getAbsolutePath).getOrElse("")

  def getIsCaretAction: Boolean = {
    caretAction
  }

  def setIsCaretAction(isCaretAction: Boolean) {
    this.caretAction = isCaretAction
  }

  def setIsSoundsPlaying(isSoundsPlaying: Boolean) {
    this.isSoundsPlaying = isSoundsPlaying
  }

  def getIsSoundsPlaying = isSoundsPlaying

  def getBamLife = bamLife

  def setBamLife(l: Long) {
    bamLife = l
  }

  def setIsBamEnabled(b: Boolean) {
    isBamEnabled = b
  }

  def getIsBamEnabled = isBamEnabled

  var heatupThreshold: Double = 0.0

  def getHeatupThreshold: Int = {
    (heatupThreshold * 100.0).toInt
  }

  def setHeatupThreshold(t: Int) {
    heatupThreshold = t / 100.0
  }

  def getIsPowerIndicatorEnabled: Boolean = {
    return powerIndicatorEnabled
  }

  def setIsPowerIndicatorEnabled(enabled: Boolean) {
    powerIndicatorEnabled = enabled
  }

  var _isCustomFlameImages: Boolean = false

  def isCustomFlameImages = _isCustomFlameImages

  def setCustomFlameImages(s: Boolean) {
    _isCustomFlameImages = s
  }


  var _isCustomBamImages: Boolean = false

  def isCustomBamImages = _isCustomBamImages

  def setCustomBamImages(s: Boolean) {
    _isCustomBamImages = s
  }

  var customFlameImageFolder = Option.empty[File]

  def setCustomFlameImageFolder(file: String) {
    customFlameImageFolder = Option(new File(file))
  }

  def getCustomFlameImageFolder: String = customFlameImageFolder.map(_.getAbsolutePath).getOrElse("")

  var customBamImageFolder = Option.empty[File]

  def setCustomBamImageFolder(file: String) {
    customBamImageFolder = Option(new File(file))
  }

  def getCustomBamImageFolder = customBamImageFolder.map(_.getAbsolutePath).getOrElse("")

}

File: src/com/nmeylan/powermode/PowerMode.java
package com.nmeylan.powermode;

import com.intellij.openapi.actionSystem.DataContext;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.components.ApplicationComponent;
import com.intellij.openapi.components.PersistentStateComponent;
import com.intellij.openapi.components.State;
import com.intellij.openapi.components.Storage;
import com.intellij.openapi.editor.EditorFactory;
import com.intellij.openapi.editor.actionSystem.EditorActionManager;
import com.intellij.util.xmlb.XmlSerializerUtil;
import com.nmeylan.powermode.color.ColorEdges;
import com.nmeylan.powermode.listeners.MyCaretListener;
import com.nmeylan.powermode.listeners.MyTypedActionHandler;
import com.nmeylan.powermode.management.ElementOfPowerContainerManager;
import org.apache.log4j.Logger;
import org.jetbrains.annotations.Nullable;

import javax.swing.KeyStroke;
import java.awt.event.InputEvent;
import java.io.File;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;


@State(name = "PowerModeII", storages = {@Storage(file = "$APP_CONFIG$/power.mode.ii.xml")})
public class PowerMode implements PersistentStateComponent<PowerMode>, ApplicationComponent {
  private static List<Integer> HOT_INPUTS = Arrays.asList(InputEvent.CTRL_DOWN_MASK, InputEvent.ALT_DOWN_MASK, InputEvent.SHIFT_DOWN_MASK);
  private boolean hotkeyHeatup = true;
  private long bamLife = 1000;
  private double gravityFactor = 21.21;
  private double sparkVelocityFactor = 4.36;
  private int sparkSize = 3;
  private int frameRate = 30;
  private int maxFlameSize = 100;
  private int maxFlameLife = 2000;
  private int heatupTime = 10000;
  private Map<KeyStroke, Long> lastKeys = new HashMap<>();

  private int keyStrokesPerMinute = 300;
  private double heatupFactor = 1.0;
  private int sparkLife = 3000;
  private int sparkCount = 10;
  private int shakeRange = 4;

  private boolean enabled = true;
  private boolean caretActionEnabled = true;
  private boolean shakeEnabled = true;
  private boolean bamEnabled = true;
  private boolean flamesEnabled = true;
  private boolean sparksEnabled = true;
  private boolean isCustomFlameImages = false;
  private boolean isCustomBamImages = false;

  private int redFrom = 200;
  private int redTo = 255;
  private int greenFrom = 0;
  private int greenTo = 255;
  private int blueFrom = 0;
  private int blueTo = 103;
  private int colorAlpha = 164;
  private double heatupThreshold = 0.0;
  private double hotkeyWeight = keyStrokesPerMinute * 0.05;

  private Optional<ElementOfPowerContainerManager> maybeElementOfPowerContainerManager = Optional.empty();

  private Optional<File> customFlameImageFolder = Optional.empty();
  private Optional<File> customBamImageFolder = Optional.empty();

  public static Logger logger() {
    return Logger.getLogger(PowerMode.class);
  }

  @Nullable
  public static PowerMode getInstance() {
    try {
      return ApplicationManager.getApplication().getComponent(PowerMode.class);
    } catch (Throwable e) {
      logger().debug("error getting component: " + e.getMessage(), e);
    }
    return null;
  }

  public ColorEdges obtainColorEdges(){
    ColorEdges edges = new ColorEdges();
    edges.setAlpha(getColorAlpha());
    edges.setRedFrom(getRedFrom());
    edges.setRedTo(getRedTo());
    edges.setGreenFrom(getGreenFrom());
    edges.setGreenTo(getGreenTo());
    edges.setBlueFrom(getBlueFrom());
    edges.setBlueTo(getBlueTo());
    return edges;
  }

  public Optional<File> flameImageFolder() {
    return isCustomFlameImages ? customFlameImageFolder : Optional.of(new File("fire/animated/256"));
  }

  public Optional<File> bamImageFolder() {
    return isCustomBamImages ? customBamImageFolder : Optional.of(new File("bam"));
  }

  public void increaseHeatup(Optional<DataContext> dataContext, KeyStroke keyStroke) {
    if (keyStroke != null) {
      long ct = System.currentTimeMillis();
      lastKeys.put(keyStroke, ct);
    }
  }

  public void reduceHeatup() {
    long ct = System.currentTimeMillis();
    lastKeys = filterLastKeys(ct);
  }

  private Map<KeyStroke, Long> filterLastKeys(Long ct) {
    return lastKeys.entrySet().stream().filter(e -> e.getValue() >= ct - heatupTime).collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
  }

  public double rawValueFactor() {
    double base = heatupFactor + ((1 - heatupFactor) * timeFactor(true));
    return (base - heatupThreshold) / (1 - heatupThreshold);
  }

  public double valueFactor() {
    double base = heatupFactor + ((1 - heatupFactor) * timeFactor(false));
    double elems = (base - heatupThreshold) / (1 - heatupThreshold);
    double max = Math.max(elems, 0.0);
    assert (max <= 1);
    assert (max >= 0);
    return max;
  }

  private double timeFactor(boolean isRaw) {
    if (heatupTime < 1000) {
      return 1;
    } else if (!lastKeys.keySet().isEmpty()) {
      double d = heatupTime / (60000.0 / keyStrokesPerMinute);
      double keysWorth = lastKeys.keySet().stream().filter(keystroke -> HOT_INPUTS.contains(keystroke.getModifiers())).count() * hotkeyWeight;
      if (isRaw) {
        return keysWorth / d;
      }
      return Math.min(keysWorth, d) / d;
    }
    return 0.0;
  }


  @Override
  public void initComponent() {
    PowerMode.logger().debug("initComponent...");
    EditorFactory editorFactory = EditorFactory.getInstance();
    maybeElementOfPowerContainerManager = Optional.of(new ElementOfPowerContainerManager());
    maybeElementOfPowerContainerManager.ifPresent(e -> editorFactory.addEditorFactoryListener(e, () -> {
    }));
    EditorActionManager editorActionManager = EditorActionManager.getInstance();
    EditorFactory
      .getInstance()
      .getEventMulticaster()
      .addCaretListener(new MyCaretListener());
    editorActionManager.getTypedAction().setupRawHandler(
      new MyTypedActionHandler(
        editorActionManager.getTypedAction().getRawHandler()));

    PowerMode.logger().debug("initComponent done");
  }

  @Override
  public void disposeComponent() {
    maybeElementOfPowerContainerManager.ifPresent(e -> e.dispose());
  }

  @Override
  public String getComponentName() {
    return "PowerModeIII";
  }

  @Override
  public PowerMode getState() {
    return this;
  }

  @Override
  public void loadState(PowerMode state) {
    XmlSerializerUtil.copyBean(state, this);
  }

  public boolean isHotkeyHeatup() {
    return hotkeyHeatup;
  }

  public void setHotkeyHeatup(boolean hotkeyHeatup) {
    this.hotkeyHeatup = hotkeyHeatup;
  }

  public long getBamLife() {
    return bamLife;
  }

  public void setBamLife(long bamLife) {
    this.bamLife = bamLife;
  }

  public double getGravityFactor() {
    return gravityFactor;
  }

  public void setGravityFactor(double gravityFactor) {
    this.gravityFactor = gravityFactor;
  }

  public double getSparkVelocityFactor() {
    return sparkVelocityFactor;
  }

  public void setSparkVelocityFactor(double sparkVelocityFactor) {
    this.sparkVelocityFactor = sparkVelocityFactor;
  }

  public int getSparkSize() {
    return sparkSize;
  }

  public void setSparkSize(int sparkSize) {
    this.sparkSize = sparkSize;
  }

  public int getFrameRate() {
    return frameRate;
  }

  public void setFrameRate(int frameRate) {
    this.frameRate = frameRate;
  }

  public int getMaxFlameSize() {
    return maxFlameSize;
  }

  public void setMaxFlameSize(int maxFlameSize) {
    this.maxFlameSize = maxFlameSize;
  }

  public int getMaxFlameLife() {
    return maxFlameLife;
  }

  public void setMaxFlameLife(int maxFlameLife) {
    this.maxFlameLife = maxFlameLife;
  }

  public int getHeatupTime() {
    return heatupTime;
  }

  public void setHeatupTime(int heatupTime) {
    this.heatupTime = Math.max(0, heatupTime);
  }

  public Map<KeyStroke, Long> getLastKeys() {
    return lastKeys;
  }

  public void setLastKeys(Map<KeyStroke, Long> lastKeys) {
    this.lastKeys = lastKeys;
  }

  public int getKeyStrokesPerMinute() {
    return keyStrokesPerMinute;
  }

  public void setKeyStrokesPerMinute(int keyStrokesPerMinute) {
    this.keyStrokesPerMinute = keyStrokesPerMinute;
  }

  public int getHeatup() {
    return (int) (heatupFactor * 100);
  }

  public void setHeatup(int heatup) {
    setHeatupFactor(heatup / 100);
  }

  public void setHeatupFactor(double heatupFactor) {
    this.heatupFactor = heatupFactor;
  }

  public int getSparkLife() {
    return sparkLife;
  }

  public void setSparkLife(int sparkLife) {
    this.sparkLife = sparkLife;
  }

  public int getSparkCount() {
    return sparkCount;
  }

  public void setSparkCount(int sparkCount) {
    this.sparkCount = sparkCount;
  }

  public int getShakeRange() {
    return shakeRange;
  }

  public void setShakeRange(int shakeRange) {
    this.shakeRange = shakeRange;
  }

  public boolean isEnabled() {
    return enabled;
  }

  public void setEnabled(boolean enabled) {
    this.enabled = enabled;
  }

  public boolean isShakeEnabled() {
    return shakeEnabled;
  }

  public void setShakeEnabled(boolean shakeEnabled) {
    this.shakeEnabled = shakeEnabled;
  }

  public boolean isBamEnabled() {
    return bamEnabled;
  }

  public void setBamEnabled(boolean bamEnabled) {
    this.bamEnabled = bamEnabled;
  }

  public boolean isFlamesEnabled() {
    return flamesEnabled;
  }

  public void setFlamesEnabled(boolean flamesEnabled) {
    this.flamesEnabled = flamesEnabled;
  }

  public boolean isSparksEnabled() {
    return sparksEnabled;
  }

  public void setSparksEnabled(boolean sparksEnabled) {
    this.sparksEnabled = sparksEnabled;
  }

  public boolean isCustomFlameImages() {
    return isCustomFlameImages;
  }

  public void setCustomFlameImages(boolean customFlameImages) {
    isCustomFlameImages = customFlameImages;
  }

  public boolean isCustomBamImages() {
    return isCustomBamImages;
  }

  public void setCustomBamImages(boolean customBamImages) {
    isCustomBamImages = customBamImages;
  }

  public int getRedFrom() {
    return redFrom;
  }

  public void setRedFrom(int redFrom) {
    if (redFrom <= redTo) {
      this.redFrom = redFrom;
    }
  }

  public int getRedTo() {
    return redTo;
  }

  public void setRedTo(int redTo) {
    if (redTo >= redFrom) {
      this.redTo = redTo;
    }
  }

  public int getGreenFrom() {
    return greenFrom;
  }

  public void setGreenFrom(int greenFrom) {
    if (greenFrom <= greenTo) {
      this.greenFrom = greenFrom;
    }
  }

  public int getGreenTo() {
    return greenTo;
  }

  public void setGreenTo(int greenTo) {
    if (greenTo >= greenFrom) {
      this.greenTo = greenTo;
    }
  }

  public int getBlueFrom() {
    return blueFrom;
  }

  public void setBlueFrom(int blueFrom) {
    if (blueFrom <= blueTo) {
      this.blueFrom = blueFrom;
    }
  }

  public int getBlueTo() {
    return blueTo;
  }

  public void setBlueTo(int blueTo) {
    if (blueTo >= blueFrom) {
      this.blueTo = blueTo;
    }
  }

  public int getColorAlpha() {
    return colorAlpha;
  }

  public void setColorAlpha(int colorAlpha) {
    this.colorAlpha = colorAlpha;
  }

  public int getHeatupThreshold() {
    return (int)(heatupThreshold * 100);
  }

  public void setHeatupThreshold(int heatupThreshold) {
    this.heatupThreshold = heatupThreshold / 100;
  }

  public Optional<ElementOfPowerContainerManager> getMaybeElementOfPowerContainerManager() {
    return maybeElementOfPowerContainerManager;
  }

  public void setMaybeElementOfPowerContainerManager(Optional<ElementOfPowerContainerManager> maybeElementOfPowerContainerManager) {
    this.maybeElementOfPowerContainerManager = maybeElementOfPowerContainerManager;
  }

  public String getCustomFlameImageFolder() {
    return customFlameImageFolder.map(f -> f.getAbsolutePath()).orElse("");
  }

  public void setCustomFlameImageFolder(String customFlameImageFolder) {
    this.customFlameImageFolder = Optional.of(new File(customFlameImageFolder));
  }

  public String getCustomBamImageFolder() {
    return customBamImageFolder.map(f -> f.getAbsolutePath()).orElse("");
  }

  public void setCustomBamImageFolder(String customBamImageFolder) {
    this.customBamImageFolder = Optional.of(new File(customBamImageFolder));
  }

  public boolean isCaretActionEnabled() {
    return caretActionEnabled;
  }

  public void setCaretActionEnabled(boolean caretActionEnabled) {
    this.caretActionEnabled = caretActionEnabled;
  }
}
